server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -50,187 +52,267 @@
+    final DruidCluster druidCluster = params.getDruidCluster();
 
     // Make sure we have enough loaded replicants in the correct tiers in the cluster before doing anything
-    for (Integer leftToLoad : loadStatus.values()) {
-      if (leftToLoad > 0) {
-        return stats;
+    for (final Object2IntMap.Entry<String> entry : targetReplicants.object2IntEntrySet()) {
+      final String tier = entry.getKey();
+      // if there are replicants loading in cluster
+      if (druidCluster.hasTier(tier) && entry.getIntValue() > currentReplicants.getOrDefault(tier, 0)) {
+        return;
       }
     }
 
-    // Find all instances of this segment across tiers
-    Map<String, Integer> replicantsByTier = params.getSegmentReplicantLookup().getClusterTiers(segment.getIdentifier());
-
-    for (Map.Entry<String, Integer> entry : replicantsByTier.entrySet()) {
+    for (final Object2IntMap.Entry<String> entry : currentReplicants.object2IntEntrySet()) {
       final String tier = entry.getKey();
-      int loadedNumReplicantsForTier = entry.getValue();
-      int expectedNumReplicantsForTier = getNumReplicants(tier);
 
-      stats.addToTieredStat(DROPPED_COUNT, tier, 0);
+      final MinMaxPriorityQueue<ServerHolder> holders = druidCluster.getHistoricalsByTier(tier);
 
-      MinMaxPriorityQueue<ServerHolder> serverQueue = params.getDruidCluster().getHistoricalsByTier(tier);
-      if (serverQueue == null) {
-        log.makeAlert("No holders found for tier[%s]", entry.getKey()).emit();
-        continue;
+      final int numDropped;
+      if (holders == null) {
+        log.makeAlert("No holders found for tier[%s]", tier).emit();
+        numDropped = 0;
+      } else {
+        final int numToDrop = entry.getIntValue() - targetReplicants.getOrDefault(tier, 0);
+        numDropped = dropForTier(numToDrop, holders, segment);
       }
 
-      List<ServerHolder> droppedServers = Lists.newArrayList();
-      while (loadedNumReplicantsForTier > expectedNumReplicantsForTier) {
-        final ServerHolder holder = serverQueue.pollLast();
-        if (holder == null) {
-          log.warn("Wtf, holder was null?  I have no servers serving [%s]?", segment.getIdentifier());
-          break;
-        }
-
-        if (holder.isServingSegment(segment)) {
-          holder.getPeon().dropSegment(
-              segment,
-              null
-          );
-          --loadedNumReplicantsForTier;
-          stats.addToTieredStat(DROPPED_COUNT, tier, 1);
-        }
-        droppedServers.add(holder);
+      stats.addToTieredStat(DROPPED_COUNT, tier, numDropped);
+    }
+  }
+
+  private static int dropForTier(
+      final int numToDrop,
+      final MinMaxPriorityQueue<ServerHolder> holders,
+      final DataSegment segment
+  )
+  {
+    int numDropped = 0;
+
+    final List<ServerHolder> droppedHolders = new LinkedList<>();
+    while (numDropped < numToDrop) {
+      final ServerHolder holder = holders.pollLast();
```

Comment:
Why need to remove holders, and then add back? Couldn't the same logic be expressed merely with iteration?

Reply:
This is because we are using `pollLast`, an `iterator` will not give us the same order (i.e., in descending available size).

Reply:
I inspected all usages of `MinMaxPriorityQueue<ServerHolder>` in the coordinator code, and it seems to me that `poll()` or `peek()` or `pollFirst()` or `peekFirst()` is never called on those min-max queues. There are only iterated "directly", elements added to them, and here, min-max is "iterated" in reverse order, via calling `pollLast()` and then adding elements back.

It means that either
 - Min-max queue is not needed, it could be a simple `PriorityQueue` in the reverse order from what is used now to create min-max queues.
 - Or, if some particular order is expected when those queues are iterated, usage of min-max queues is a mistake, because it doesn't guarantee any particular iteration order. In this case, it should be replaced with `TreeSet`.

Also @fjy could you please comment here

Reply:
@leventov @fjy From its use in this case, it seems the order is required. Should I modify the usages of `MinMaxPriorityQueue` to `TreeSet` then?